WordPress/WordPressShareExtension/String+Extensions.swift

```diff
@@ -88,6 +88,13 @@ extension String {
         return self + returnURLString
     }
 
+    /// Returns a Boolean value indicating if this String begins with provided prefix. The comparison is both case-insensitive and diacritic-insensitive.
+    /// - Parameter prefix: The provideded prefix to verify.
+    /// - Returns: true if this string begins with the provided prefix, false otherwise.
+    func beginsWith(prefix: String) -> Bool {
+        return range(of: prefix, options: [.anchored, .caseInsensitive, .diacriticInsensitive]) != nil
+    }
+
```

Comment:
Given this specifically does a case- and diacritics-insensitive test, which is what makes it differ from Foundation's built-in `String.hasPrefix`, I'd suggest naming this method accordingly, to avoid other devs starting to see "oh, autocomplete suggests me a `beginsWith(prefix:)` method on `String`, I'll use it for my feature" without realizing it's ignoring case and diacritics.

 - Maybe `hasCaseAndDiacriticsInsensitivePrefix` (though this is a bit of a mouthful)?
 - Or `hasFoldedPrefix` (with "folded" being used here in the sense if "String folding" as in [here](https://developer.apple.com/documentation/foundation/nsstring/1413779-folding)… though this term of art might not be widely known so maybe not super clear either…
 - Feel free to suggest anything else. Naming is hard!

Reply:
Hi @AliSoftware , I see your point. I don't like a big name for this function, either. What if we keep the name `hasPrefix` and expose the options as a parameter allowing the user to define the compare options?

```suggestion
    /// Returns a Boolean value indicating if this String begins with provided prefix.
    /// - Parameter prefix: The provided prefix to verify.
    /// - Parameter options: The provided compare options.
    /// - Returns: true if this String begins with the provided prefix following with the given compare options, false otherwise.
    func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
        return range(of: prefix, options: options) != nil
    }

```

Reply:
I like it 👍

Reply:
This new way is a bit risky. If someone doesn't pass `.anchored` to this function, it will lose its definition as it will no longer be looking for prefixes. I think we'd have to make the API more strict than this to avoid misuse. Wdyt?

Reply:
Nice catch and very good point @hassaanelgarem . What about keeping the method name and signature as `hasPrefix(_ prefix: String, with options: CompareOptions)` but make sure the implementation adds `.anchored` to the `options` set before using it in the `range(…)` call?

Reply:
Hey @hassaanelgarem, what if we use a `preconditionFailure` call along with @AliSoftware's suggestion?
```swift
    /// Returns a Boolean value indicating if this String begins with provided prefix.
    /// - Parameter prefix: The provided prefix to verify.
    /// - Parameter options: The provided compare options. It must contain CompareOption.anchored.
    /// - Returns: true if this String begins with the provided prefix following the given compare options, false otherwise.
    func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
        if !options.contains(.anchored) {
            preconditionFailure("options must contain CompareOption.anchored")
        }
        return range(of: prefix, options: options) != nil
    }
```

Reply:
`preconditionFailure` not only will only be caught at runtime, but will also put the responsibility to the caller to have to specify an option they shouldn't have to worry about because they shouldn't have to know about the implementation detail of your `hasPrefix` method (ie the fact that it uses `range(of:…)` under the hood — as opposed to some other possible implementation you could have used here, like `self.folding(options: options).hasPrefix(prefix.folding(options: options))` for example, which would have been equally valid without requiring the caller to provide `.anchored` as part of the options at call site.

So TL;DR: it shouldn't be the caller's responsibility to add `.anchored` to the options. The need for `.anchored` is an implementation detail, not a caller requirement.

Reply:
Yes, that makes total sense @AliSoftware. So, let me go back to your original suggestion:
```swift
    func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
        var optionsSet = options
        if !optionsSet.contains(.anchored) {
            optionsSet.update(with: .anchored)
        }
        return range(of: prefix, options: optionsSet) != nil
    }
```
@hassaanelgarem wdyt?

Reply:
Since `CompareOptions` is an `OptionSet` (so a set and not an `Array`) I think you can add the `.anchored` flag to the set unconditionally.

func hasPrefix(_ prefix: String, with options: CompareOptions) -> Bool {
        let fullOptions = options.union([.anchored])
        return range(of: prefix, options: fullOptions) != nil
    }

Reply:
@igoriols I agree with @AliSoftware's suggestion 👌 
It's a neat solution and ensures the function is used correctly.

Reply:
Thanks, @hassaanelgarem and @AliSoftware! I applied the changes you suggested.